Sources/Leaf/Parser.swift

```diff
@@ -0,0 +1,347 @@
+            return nil
+        }
+
+        let syntax: Syntax
+
+        switch byte {
+        case .numberSign:
+            syntax = try extractTag()
+        default:
+            let start = scanner.offset
+            let line = scanner.line
+            let column = scanner.column
+
+            let bytes = try extractRaw()
+
+            let source = Source(line: line, column: column, range: start..<scanner.offset)
+            return Syntax(kind: .raw(data: bytes), source: source)
+        }
+
+        return syntax
+    }
+
+    private func extractTag() throws -> Syntax {
+        let start = scanner.offset
+        let line = scanner.line
+        let column = scanner.column
+
+        try expect(.numberSign)
+        let id = try extractIdentifier()
+        let params = try extractParameters()
+        try skipWhitespace()
+
+        let body: [Syntax]?
+
+        if let byte = scanner.peek() {
+            if byte == .leftCurlyBracket {
+                body = try extractBody()
+            } else {
+                body = nil
+            }
+        } else {
+            body = nil
+        }
+
+        let kind: SyntaxKind = .tag(
+            name: id,
+            parameters: params,
+            body: body
+        )
+        
+        let source = Source(line: line, column: column, range: start..<scanner.offset)
+        return Syntax(kind: kind, source: source)
+    }
+
+    private func extractBody() throws -> [Syntax] {
+        try expect(.leftCurlyBracket)
+        let body = try bytes(until: .rightCurlyBracket)
+        try expect(.rightCurlyBracket)
+        let parser = Parser(body)
+        return try parser.parse()
+    }
+
+    private func extractRaw() throws -> Bytes {
+        return try bytes(until: .numberSign)
+    }
+
+    private func bytes(until: Byte) throws -> Bytes {
+        var previous: Byte?
+
+        var bytes: Bytes = []
+        while let byte = scanner.peek(), byte != until || previous == .backSlash {
+            try scanner.pop()
+            if byte != until && previous == .backSlash {
+                bytes.append(.backSlash)
+            }
+            if byte != .backSlash {
+                bytes.append(byte)
+            }
+            previous = byte
+        }
+
+        return bytes
+    }
+
+    private func extractIdentifier() throws -> Syntax {
+        let start = scanner.offset
+        let line = scanner.line
+        let column = scanner.column
+
+        while let byte = scanner.peek(), byte.isAlphanumeric {
+           try scanner.pop()
+        }
+        
+        let bytes = scanner.bytes[start..<scanner.offset]
+
+        let kind: SyntaxKind = .identifier(name: bytes.makeString())
+        let source = Source(line: line, column: column, range: start..<scanner.offset)
+
+        return Syntax(kind: kind, source: source)
+    }
+
+    private func extractParameters() throws -> [Syntax] {
+        try expect(.leftParenthesis)
```

Comment:
This is most likely intended, but just writing a warning here. This is whitespace sensitive. `#example ()` will fail.